src/statusicon.cpp

```diff
@@ -43,117 +46,122 @@ constexpr const char* LOGO_GENERIC_ON =
-      setIndicatorColor(RED_COLOR);
-      break;
-    default:
-      logger.error() << "Unhandled status indicator for connection stability";
-      setIndicatorColor(QColor());
-  }
-}
-
-const QIcon& StatusIcon::icon() {
-  if (m_icon.isNull()) {
-    stateChanged();
-    Q_ASSERT(!m_icon.isNull());
-  }
-
-  return m_icon;
-}
-
-void StatusIcon::stateChanged() {
-  logger.debug() << "State changed";
-
-  m_animatedIconTimer.stop();
+const QString StatusIcon::iconString() {
+  logger.debug() << "Icon string";
 
   MozillaVPN* vpn = MozillaVPN::instance();
   Q_ASSERT(vpn);
 
   // If we are in a non-main state, we don't need to show special icons.
   if (vpn->state() != MozillaVPN::StateMain) {
-    setIcon(LOGO_GENERIC, false);
-    setIndicatorColor(QColor());
-    return;
+    return LOGO_GENERIC;
   }
-
   switch (vpn->controller()->state()) {
-#if defined(MVPN_LINUX) || defined(MVPN_WINDOWS)
-    case Controller::StateOn:
-      [[fallthrough]];
-    case Controller::StateOff:
-      setIcon(LOGO_GENERIC, true);
-      break;
-#else
     case Controller::StateOn:
-      setIcon(LOGO_GENERIC_ON, false);
-      // Make sure the indicator color is set initially
-      setIndicatorColor(GREEN_COLOR);
+      m_animatedIconTimer.stop();
+      return LOGO_GENERIC_ON;
       break;
     case Controller::StateOff:
-      setIcon(LOGO_GENERIC_OFF, false);
-      setIndicatorColor(QColor());
+      m_animatedIconTimer.stop();
+      return LOGO_GENERIC_OFF;
       break;
-#endif
     case Controller::StateSwitching:
+      logger.debug() << "lel StateSwitching";
       [[fallthrough]];
     case Controller::StateConnecting:
+      logger.debug() << "lel StateConnectin";
       [[fallthrough]];
     case Controller::StateConfirming:
+      logger.debug() << "lel StateConfirmin";
       [[fallthrough]];
     case Controller::StateDisconnecting:
-      activateAnimation();
+      if (!m_animatedIconTimer.isActive()) {
+        activateAnimation();
+      }
+      return ANIMATED_LOGO_STEPS[m_animatedIconIndex];
       break;
     default:
-      setIcon(LOGO_GENERIC, false);
+      m_animatedIconTimer.stop();
+      return LOGO_GENERIC;
       break;
   }
 }
 
-QIcon StatusIcon::drawStatusIndicator(const QString& iconUrl) const {
+const QColor& StatusIcon::indicatorColor() {
+  logger.debug() << "Set color";
+
+  MozillaVPN* vpn = MozillaVPN::instance();
+  Q_ASSERT(vpn);
+
+  if (vpn->state() != MozillaVPN::StateMain ||
+      vpn->controller()->state() != Controller::StateOn) {
+    return INVALID_COLOR;
+  }
+
+  switch (vpn->connectionHealth()->stability()) {
+    case ConnectionHealth::Stable:
+      return GREEN_COLOR;
+    case ConnectionHealth::Unstable:
+      return ORANGE_COLOR;
+    case ConnectionHealth::NoSignal:
+      return RED_COLOR;
+    default:
+      logger.error() << "Unhandled status indicator for connection stability";
+      return INVALID_COLOR;
+  }
+
+  return m_indicatorColor;
+}
+
+void StatusIcon::stateChanged() {
```

Comment:
What about if we call this method `refreshNeeded` or `resetIcon`? Asking because it's not connected to state changing anymore.

Reply:
Sounds good!